### PR TITLE
[INSIGHTOCP-1965] Add pre-commit linters to GitHub CI workflow

### PR DESCRIPTION
**Pull request by jholecek-rh**
See individual commits for details.

### CODE REVIEW

**Review by ikerreyes — COMMENTED** [review]
Not sure if I am following the rationale for removing `validate.py` for pre-commit.

> The pre-commit hook definition for valdiate.py didn't use requirements.txt to specify validate.py dependencies

I agree it should be treated in a different way, but I do not think that invalidates the point of running this in pre-commit.

> The CI pipeline is defined using GitHub workflows. We do not need to provide a single command that executes the whole CI pipeline, not even for local execution.

What is the relation with `validate.py`?  That script is only about the schema validation, nothing to do with other linters, or am I missing something?

**Comment by jholecek-rh** [comment]
@ikerreyes 

Yeah, I had a feeling I didn't explain it well. I tried to keep it short. Too short it seems. Let me try again.

> What is the relation with `validate.py`?

I don't understand what you mean. The relation between `validate.py` and what?

> That script is only about the schema validation, nothing to do with other linters, or am I missing something?

The current repo structure is not sustainable and the current CI pipeline has gaps. I don't know if `validate.py` will remain and what its scope would be when we implement all the required changes. Will we want validation to duplicate steps of the build process independently? I thought that having `validate.py` as a separate documented step would make it more visible for the upcoming discussions about these topics. 

> I do not think that invalidates the point of running this in pre-commit.

I would be curious to know what benefits you see in running `validate.py` as part of `pre-commit`. Here's how I came to my proposed solution.

I've got these three requirements for the CI pipeline:

1. I want the CI pipeline to run all checks that devs might be asked to run locally (or as many of them as feasible); people would forget eventually.
2. I want devs to be able to run any CI pipeline steps locally; useful for troubleshooting.
3. I want the documentation for running the checks locally to follow the CI pipeline; it is easy to maintain consistency between the two that way.

The original documentation differed from the CI pipeline in `pre-commit`. The pipeline didn't apply the `trailing-whitespace` and `end-of-file-fixer` hooks. The existing files didn't comply with the two hooks already.  I had to choose: add `pre-commit` to the pipeline or drop it completely. I chose to add it, because I consider the two hooks useful and I see potential for more hooks in the future.

I didn't want the CI pipeline to execute `validate.py` twice, so I had to choose whether to define the step in `pre-commit` or the CI pipeline directly. I opted for the pipeline because:

1. We need a Python environment for `pre-commit` anyway. That means a documented way to set up  the environment for my requirement 2 above.
2. It would be weird if we couldn't run `python validate.py` in that environment, so the environment needs to include `validate.py` dependencies.
3. I didn't want to maintain `validate.py` dependencies at two different places.

**Comment by ikerreyes** [comment]
@jholecek-rh My bad. I misunderstood the intention of the changes.

> I thought that having validate.py as a separate documented step would make it more visible for the upcoming discussions about these topics.

Yes, it is good point, specially if you have concerns about its use long term. Makes sense.

> I chose to add it, because I consider the two hooks useful and I see potential for more hooks in the future.

Yes, I agree.

> I would be curious to know what benefits you see in running validate.py as part of pre-commit.

It provides automation directly on devs machine (if they install the hooks) rather than waiting for the CI to run in the repo upstream, so feedback is faster. The second benefit is that it encapsulates dependencies. This is a bit related to the point 2 you make to opt for the pipeline option. Development env should contain those dependencies if we want to run it as `python validate`, but if we just want them as pre-commit hooks, then we do not need them in the `requirements.txt`.

Said that, if 

> It would be weird if we couldn't run python validate.py

then I am fine to remove it from the `pre-commit` hook.

All this said, then I am OK with the changes, I just didn't understood the context fully, so thank you for taking the time to explain to me.

Just for completeness

> > What is the relation with validate.py?

> I don't understand what you mean. The relation between validate.py and what?

I was referring to the relation between `validate.py` and the single command for running the whole pipeline. But I just misunderstood your comment, my bad. Ignore my question.

**Review by ikerreyes — APPROVED** [review]
